File: app/src/main/scala/com/scala/examples/companion/Person.scala
package com.scala.examples.companion

// Define a class with a private constructor
class Person private(val name: String, val age: Int)

// Define the companion object
object Person {
  // Factory method to create a Person instance
  def apply(name: String, age: Int): Person = new Person(name, age)
}

// Main application
object MainApp extends App {
  // Use the factory method in the companion object to create an instance
  val person = Person("Alice", 30)
  println(s"Name: ${person.name}, Age: ${person.age}")
}

File: app/src/main/java/examples/companion/Person.java
package examples.companion;

public class Person {
    private String name;
    private int age;

    // Private constructor to prevent direct instantiation
    private Person(String name, int age) {
        this.name = name;
        this.age = age;
    }

    // Static factory method
    public static Person createPerson(String name, int age) {
        return new Person(name, age);
    }

    public String getName() {
        return name;
    }

    public int getAge() {
        return age;
    }
}

class Main {
    public static void main(String[] args) {
        Person person = Person.createPerson("Alice", 30);
        System.out.println("Name: " + person.getName() + ", Age: " + person.getAge());
    }
}
